google/cloud/pubsub/internal/blocking_publisher_tracing_connection_test.cc

```diff
@@ -0,0 +1,196 @@
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+#include "google/cloud/pubsub/internal/blocking_publisher_tracing_connection.h"
+#include "google/cloud/pubsub/blocking_publisher_connection.h"
+#include "google/cloud/pubsub/message.h"
+#include "google/cloud/pubsub/mocks/mock_blocking_publisher_connection.h"
+#include "google/cloud/pubsub/topic.h"
+#include "google/cloud/internal/make_status.h"
+#include "google/cloud/status.h"
+#include "google/cloud/status_or.h"
+#include "google/cloud/testing_util/opentelemetry_matchers.h"
+#include "google/cloud/testing_util/status_matchers.h"
+#include <gmock/gmock.h>
+#include <opentelemetry/trace/semantic_conventions.h>
+
+namespace google {
+namespace cloud {
+namespace pubsub_internal {
+namespace {
+
+auto constexpr kErrorCode = static_cast<int>(StatusCode::kAborted);
+
+using ::google::cloud::pubsub::MessageBuilder;
+using ::google::cloud::pubsub::Topic;
+using ::google::cloud::pubsub_internal::MakeBlockingPublisherTracingConnection;
+using ::google::cloud::pubsub_mocks::MockBlockingPublisherConnection;
+using ::google::cloud::testing_util::InstallSpanCatcher;
+using ::google::cloud::testing_util::OTelAttribute;
+using ::google::cloud::testing_util::SpanHasAttributes;
+using ::google::cloud::testing_util::SpanHasInstrumentationScope;
+using ::google::cloud::testing_util::SpanKindIsClient;
+using ::google::cloud::testing_util::SpanKindIsProducer;
+using ::google::cloud::testing_util::SpanNamed;
+using ::google::cloud::testing_util::SpanWithStatus;
+using ::google::cloud::testing_util::StatusIs;
+using ::google::cloud::testing_util::ThereIsAnActiveSpan;
+using ::testing::_;
+using ::testing::AllOf;
+using ::testing::ElementsAre;
+using ::testing::Not;
+
+TEST(BlockingPublisherTracingConnectionTest, PublishSpanOnSuccess) {
+  namespace sc = ::opentelemetry::trace::SemanticConventions;
+  auto span_catcher = InstallSpanCatcher();
+  auto mock = std::make_shared<MockBlockingPublisherConnection>();
+  EXPECT_CALL(*mock, Publish)
+      .WillOnce([&](pubsub::BlockingPublisherConnection::PublishParams const&) {
+        EXPECT_TRUE(ThereIsAnActiveSpan());
+        return StatusOr<std::string>("test-id-0");
+      });
+  auto connection = MakeBlockingPublisherTracingConnection(std::move(mock));
+
+  Topic const topic("test-project", "test-topic");
+  auto response =
+      connection->Publish({topic, pubsub::MessageBuilder{}
+                                      .SetData("test-data-0")
+                                      .SetOrderingKey("ordering-key-0")
+                                      .Build()});
+
+  EXPECT_STATUS_OK(response);
+  EXPECT_THAT(
+      span_catcher->GetSpans(),
+      ElementsAre(AllOf(
+          SpanHasInstrumentationScope(), SpanKindIsProducer(),
+          SpanNamed("projects/test-project/topics/test-topic send"),
+          SpanWithStatus(opentelemetry::trace::StatusCode::kOk),
+          SpanHasAttributes(
+              OTelAttribute<std::string>(sc::kMessagingSystem, "pubsub"),
```

Comment:
for a future PR: it seems that "pubsub" is not unique enough, the existing examples in:

https://opentelemetry.io/docs/specs/semconv/messaging/messaging-spans/#messaging-attributes

suggest it should be something like Google Pub/Sub.  Extra credit: work with the OTel and service folks to define a standard name and attributes for Cloud Pub/Sub:

https://opentelemetry.io/docs/specs/semconv/messaging/

Reply:
Ack, https://github.com/googleapis/google-cloud-cpp/issues/12952